rac2: add token counter and stream metrics

kvoli opened the pull request:
This commit introduces metrics related to stream eval tokens and stream
send tokens. Hooking up these metrics to the registry will be in a
subsequent commit.

There are two separate metric structs used:

1. `tokenCounterMetrics`, which only contains counter and is shared
   among all `tokenCounter`s on the same node. Each `tokenCounter`
   updates the shared counters after `adjust` is called.
2. `tokenStreamMetrics`, which is updated periodically by calling
   `UpdateMetricGauges` via the `StreamTokenCounterProvider`, which is
   one per node.

Metrics related to `WaitForEval` (as well as blocked stream logging) are
also deferred to a subsequent commit.

Part of: https://github.com/cockroachdb/cockroach/issues/128031
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129350)
<!-- Reviewable:end -->

kvoli committed:
rac2: de-interface token counter and use syncutil map

Prior to this change, `TokenCounter` provided an interface implemented
by `*tokenCounter`. As there is only one implementation, de-interface
`TokenCounter`. Also, store the TokenCounter in a `syncutil.Map`, as opposed
to a native mutex protected map.

Epic: CRDB-37515
Release note: None

kvoli committed:
rac2: add token counter and stream metrics

This commit introduces metrics related to stream eval tokens and stream
send tokens. Hooking up these metrics to the registry will be in a
subsequent commit.

There are two separate metric structs used:

1. `tokenCounterMetrics`, which only contains counter and is shared
   among all `tokenCounter`s on the same node. Each `tokenCounter`
   updates the shared counters after `adjust` is called.
2. `tokenStreamMetrics`, which is updated periodically by calling
   `UpdateMetricGauges` via the `StreamTokenCounterProvider`, which is
   one per node.

Metrics related to `WaitForEval` (as well as blocked stream logging) are
also deferred to a subsequent commit.

Part of: #128031
Release note: None

kvoli commented:
Open to suggestions for the migration. I was planning on just hooking these up and having both metrics be enabled at once. We *could* reuse the existing metrics with some heavy refactoring, if that were desirable.

sumeerbhola reviewed (changes requested):
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129350)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli and @pav-kv)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/metrics.go` line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5OHXVa1hrg7Z2PmoZj:-O5OHXVa1hrg7Z2PmoZk:b-q9t5wy) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ae6f673c667bb604af6a42a1c907365583d6797/pkg/kv/kvserver/kvflowcontrol/rac2/metrics.go#L161)):*
> ```Go
> 	ElasticFlowTokensDeducted    *metric.Counter
> 	ElasticFlowTokensReturned    *metric.Counter
> 	ElasticFlowTokensUnaccounted *metric.Counter
> ```

is there a reason these are not arrays too?

btw, I am not a fan of the callback based pattern of adjustment that was used in the RACv1 code -- it is error prone and breaks abstraction boundaries. The two patterns I prefer:
- The kvserver metrics pattern, where something calls the various components periodically, extracts stats structs and updates both gauges and counters. It does require more work to implement, since the various components need to provide their cumulative stats. In cases like Pebble this was already the case.
- The data-structures are responsible for updating the (sometimes shared) cumulative metrics when they update their state. This distributes the metrics update to where the state updates are happening which IMO is desirable. Less abstraction leakage.

The other minor nit with the RACv1 structuring is that it prefixed everything with `kvadmission.flow_controller`. RACv1 was not in the kvadmission package, and some metrics relate to token counters and some relate to waiting etc. I think we need to address this since we now have both eval and send tokens and concepts like `WaitDuration` aren't necessarily relevant for send tokens, and they are relevant to WaitForEval. Also, since we have both eval and send tokens, I withdraw my earlier position that we should share metrics with RACv1. There was a valid point made yesterday by Andrew about not naming these v2.

So I'd suggest a naming scheme that segments out the TokenCounter metrics, such as
kvflowcontrol.tokens.<eval|send>.<regular|elastic>.<deducted|returned|unaccounted|available>

Then have a struct like
```
type tokenCounterMetrics {
   deducted [NumWorkClasses]*metric.Counter
   returned ...
   ...
}
```
containing all the cumulative metrics that a `TokenCounter` needs to adjust. Provide the `TokenCounter` the struct when creating it in `StreamTokenCounterProvider`. `StreamTokenCounterProvider` is the one that will interface with a registry since there is only one provider is a node.
Additionally `StreamTokenCounterProvider` will have a separate struct for gauge metrics (all functional gauges) that require iteration over the TokenCounters (e.g. stream count, available tokens).

All the `StreamTokenCounterProvider` and `TokenCounter` changes belong in one PR.

The remaining are eval wait metrics, which could all be named with the kvflowcontrol.eval_wait prefix. We could possibly update them in `RangeController.WaitForEval`. `RangeController` would expect a metrics struct to be provided to it as part of its options.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5OEA0Q3SZo1ggiNEFC:-O5OEA0R0dy6NzPp-k-L:bymhz9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ec7bdc73bde078b3548a9f5f6603b3fe93b4a4e/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L37)):*
> ```Go
> // Eval returns the evaluation token counter for the given stream.
> func (p *StreamTokenCounterProvider) Eval(stream kvflowcontrol.Stream) *TokenCounter {
> 	t, _ := p.evalCounters.LoadOrStore(stream, NewTokenCounter(p.settings))
> ```

We shouldn't be using `LoadOrStore` directly since `NewTokenCounter` is expensive (it allocates). The usual pattern is to first call `Load` and if it is not found, call `LoadOrStore`. And if we expect a high concurrency in those that fail the `Load`, and the new is very expensive, do a creation mutex acquisition, call `Load` a second time, and if it fails call `LoadOrStore` while holding the mutex. In this case I think that is not necessary.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5OEymP33vnJub3U5ud:-O5OEymP33vnJub3U5ue:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ec7bdc73bde078b3548a9f5f6603b3fe93b4a4e/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L43)):*
> ```Go
> // Send returns the send token counter for the given stream.
> func (p *StreamTokenCounterProvider) Send(stream kvflowcontrol.Stream) *TokenCounter {
> 	t, _ := p.sendCounters.LoadOrStore(stream, NewTokenCounter(p.settings))
> ```

ditto
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/token_counter.go` line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5OF37200nGnw93DfK_:-O5OF37200nGnw93DfKa:b-hi71di) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ec7bdc73bde078b3548a9f5f6603b3fe93b4a4e/pkg/kv/kvserver/kvflowcontrol/rac2/token_counter.go#L144)):*
> ```Go
> // kvflowcontrol.Stream. It's used to synchronize handoff between threads
> // returning and waiting for flow tokens.
> type TokenCounter struct {
> ```

is this exported since it is getting called outside the package (or will be)?


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129350)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv and @sumeerbhola)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/metrics.go` line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5OHXVa1hrg7Z2PmoZj:-O5OTorD4pyJ_qlIu0CX:b-5huxf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ae6f673c667bb604af6a42a1c907365583d6797/pkg/kv/kvserver/kvflowcontrol/rac2/metrics.go#L161)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is there a reason these are not arrays too?

btw, I am not a fan of the callback based pattern of adjustment that was used in the RACv1 code -- it is error prone and breaks abstraction boundaries. The two patterns I prefer:
- The kvserver metrics pattern, where something calls the various components periodically, extracts stats structs and updates both gauges and counters. It does require more work to implement, since the various components need to provide their cumulative stats. In cases like Pebble this was already the case.
- The data-structures are responsible for updating the (sometimes shared) cumulative metrics when they update their state. This distributes the metrics update to where the state updates are happening which IMO is desirable. Less abstraction leakage.

The other minor nit with the RACv1 structuring is that it prefixed everything with `kvadmission.flow_controller`. RACv1 was not in the kvadmission package, and some metrics relate to token counters and some relate to waiting etc. I think we need to address this since we now have both eval and send tokens and concepts like `WaitDuration` aren't necessarily relevant for send tokens, and they are relevant to WaitForEval. Also, since we have both eval and send tokens, I withdraw my earlier position that we should share metrics with RACv1. There was a valid point made yesterday by Andrew about not naming these v2.

So I'd suggest a naming scheme that segments out the TokenCounter metrics, such as
kvflowcontrol.tokens.<eval|send>.<regular|elastic>.<deducted|returned|unaccounted|available>

Then have a struct like
```
type tokenCounterMetrics {
   deducted [NumWorkClasses]*metric.Counter
   returned ...
   ...
}
```
containing all the cumulative metrics that a `TokenCounter` needs to adjust. Provide the `TokenCounter` the struct when creating it in `StreamTokenCounterProvider`. `StreamTokenCounterProvider` is the one that will interface with a registry since there is only one provider is a node.
Additionally `StreamTokenCounterProvider` will have a separate struct for gauge metrics (all functional gauges) that require iteration over the TokenCounters (e.g. stream count, available tokens).

All the `StreamTokenCounterProvider` and `TokenCounter` changes belong in one PR.

The remaining are eval wait metrics, which could all be named with the kvflowcontrol.eval_wait prefix. We could possibly update them in `RangeController.WaitForEval`. `RangeController` would expect a metrics struct to be provided to it as part of its options.
</blockquote></details>

TYFTR

I'll go through and do some refactoring. Agree on most of these points, these diffs were taken partially as is from the prototype which did the same thing with the existing v1 metrics.


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
Should be good for another round @sumeerbhola.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-:-O5PF5_P35klWlD9Jk3C:b4uxd8j)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv and @sumeerbhola)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/metrics.go` line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5OHXVa1hrg7Z2PmoZj:-O5PEw2LAuuFu-k_P7Uf:b-gp7ltm) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ae6f673c667bb604af6a42a1c907365583d6797/pkg/kv/kvserver/kvflowcontrol/rac2/metrics.go#L161)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

TYFTR

I'll go through and do some refactoring. Agree on most of these points, these diffs were taken partially as is from the prototype which did the same thing with the existing v1 metrics.
</blockquote></details>

Updated to slim down this PR to just the token/stream metrics. The logging, hookup and eval metrics will be in separate PRs, ontop of this.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5OEA0Q3SZo1ggiNEFC:-O5OlUyJCq3Z9bs-bAFa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ec7bdc73bde078b3548a9f5f6603b3fe93b4a4e/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L37)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

We shouldn't be using `LoadOrStore` directly since `NewTokenCounter` is expensive (it allocates). The usual pattern is to first call `Load` and if it is not found, call `LoadOrStore`. And if we expect a high concurrency in those that fail the `Load`, and the new is very expensive, do a creation mutex acquisition, call `Load` a second time, and if it fails call `LoadOrStore` while holding the mutex. In this case I think that is not necessary.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5OEymP33vnJub3U5ud:-O5OlVq6Bhb8xUljFcY4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ec7bdc73bde078b3548a9f5f6603b3fe93b4a4e/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L43)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/token_counter.go` line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5OF37200nGnw93DfK_:-O5On66b-D-qh6WmZ-Ax:b-3h52r1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ec7bdc73bde078b3548a9f5f6603b3fe93b4a4e/pkg/kv/kvserver/kvflowcontrol/rac2/token_counter.go#L144)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is this exported since it is getting called outside the package (or will be)?
</blockquote></details>

No reason it should be, changed to private.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 8 files at r3, 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129350)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli and @pav-kv)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 81 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5Q8qdGDBPvckbpqia-:-O5Q8qdGDBPvckbpqia0:bejkix5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede81ada6045619d50874b1a761e52bc776970ca/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L81)):*
> ```Go
> 		return func(stream kvflowcontrol.Stream, t *tokenCounter) bool {
> 			count[metricType][regular]++
> 			count[flowControlEvalMetricType][elastic]++
> ```

why is this not indexed by `metricType`. Same question below.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 97 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5Q8y073-In2oLhq6d8:-O5Q8y073-In2oLhq6d9:b-bfxtsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede81ada6045619d50874b1a761e52bc776970ca/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L97)):*
> ```Go
> 
> 	p.evalCounters.Range(gaugeUpdateFn(flowControlEvalMetricType))
> 	p.sendCounters.Range(gaugeUpdateFn(flowControlEvalMetricType))
> ```

shouldn't this be flowControlSendMetricType?

Worth adding some test cases that would find bugs.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/token_counter.go` line 93 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5QAEL92iQ58R5tqrv-:-O5QAEL92iQ58R5tqrv0:b5m2uhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede81ada6045619d50874b1a761e52bc776970ca/pkg/kv/kvserver/kvflowcontrol/rac2/token_counter.go#L93)):*
> ```Go
> type deltaStats struct {
> 	noTokenDuration                time.Duration
> 	tokensDeducted, tokensReturned kvflowcontrol.Tokens
> ```

we don't seem to use these `tokensDeducted` and `tokensReturned` stats. Was this carried over from v1 because of the different way it was doing cumulative metrics?
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/token_counter_test.go` line 44 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5Q7jpI4lf0AQ86-NKd:-O5Q7jpI4lf0AQ86-NKe:b-nkeu6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede81ada6045619d50874b1a761e52bc776970ca/pkg/kv/kvserver/kvflowcontrol/rac2/token_counter_test.go#L44)):*
> ```Go
> }
> 
> func TestTokenAdjustment(t *testing.T) {
> ```

is it easy to check the metrics values in a few places in the tests in this file?


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129350)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv and @sumeerbhola)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 81 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5Q8qdGDBPvckbpqia-:-O5UBH9l1Fzz6njIZkjC:b-tsh4pv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede81ada6045619d50874b1a761e52bc776970ca/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L81)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why is this not indexed by `metricType`. Same question below.
</blockquote></details>

Bug, fixed.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go` line 97 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5Q8y073-In2oLhq6d8:-O5UBIN57OE-lTO2CFXj:b-b6n7a1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede81ada6045619d50874b1a761e52bc776970ca/pkg/kv/kvserver/kvflowcontrol/rac2/store_stream.go#L97)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

shouldn't this be flowControlSendMetricType?

Worth adding some test cases that would find bugs.
</blockquote></details>

It should be. Added some tests.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/token_counter.go` line 93 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5QAEL92iQ58R5tqrv-:-O5U4QMo6sieRhuakiDc:b-mznao0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede81ada6045619d50874b1a761e52bc776970ca/pkg/kv/kvserver/kvflowcontrol/rac2/token_counter.go#L93)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

we don't seem to use these `tokensDeducted` and `tokensReturned` stats. Was this carried over from v1 because of the different way it was doing cumulative metrics?
</blockquote></details>

These will be used for blocked stream logging, I added them in as part of the deltaStats but only the noTokenDuration is being used atm. I can remove these from the PR and move to follow up w/ the logging but would prefer to keep them otherwise.
___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/token_counter_test.go` line 44 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5Q7jpI4lf0AQ86-NKd:-O5UB3EeCodbqnB4hAZx:bm2y7c4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede81ada6045619d50874b1a761e52bc776970ca/pkg/kv/kvserver/kvflowcontrol/rac2/token_counter_test.go#L44)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is it easy to check the metrics values in a few places in the tests in this file?
</blockquote></details>

Yeah pretty easy. Added.


<!-- Sent from Reviewable.io -->

kvoli commented:
Should be ready for another look @sumeerbhola.

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 8 files at r7, 7 of 7 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-:-O5YrDoG-QDdrAFmn9my:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @kvoli and @pav-kv)

___
*[`pkg/kv/kvserver/kvflowcontrol/rac2/token_counter.go` line 93 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129350#-O5QAEL92iQ58R5tqrv-:-O5YqehBBz09iqewinAd:b5k6zdf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede81ada6045619d50874b1a761e52bc776970ca/pkg/kv/kvserver/kvflowcontrol/rac2/token_counter.go#L93)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

These will be used for blocked stream logging, I added them in as part of the deltaStats but only the noTokenDuration is being used atm. I can remove these from the PR and move to follow up w/ the logging but would prefer to keep them otherwise.
</blockquote></details>

Fine to keep them


<!-- Sent from Reviewable.io -->

kvoli commented:
TYFTR!

bors r=sumeerbhola
